### PR TITLE
Feature/cc 742 monitor

### PR DESCRIPTION
**Pull request by wplu**
https://jira.zenoss.com/browse/CC-742

DEMO:
```
~/src/europa/src/golang/src/github.com/control-center/serviced/coordinator/storage
# plu@plu-9: go test -v
=== RUN TestClient
--- SKIP: TestClient (0.00 seconds)
	client_test.go:34: Test cluster is not set up properly
=== RUN TestMonitorVolume
I0124 11:18:21.416346 17580 monitor.go:96] monitoring DFS export info for DFS volume /tmp/storage199463140 at polling interval: 3s
I0124 11:18:23.416319 17580 monitor_test.go:56] ==== starting writer
I0124 11:18:23.416379 17580 monitor.go:41] writing remote monitor file to DFS volume /tmp/storage199463140 at interval: 1s
I0124 11:18:23.416404 17580 monitor.go:45] writing DFS file /tmp/storage199463140/monitor/127.0.0.1
I0124 11:18:24.416664 17580 monitor.go:45] writing DFS file /tmp/storage199463140/monitor/127.0.0.1
I0124 11:18:24.416893 17580 monitor.go:145] file /tmp/storage199463140/monitor/127.0.0.1 was modified 6.667821ms ago
I0124 11:18:24.416911 17580 monitor.go:117] remote agent 127.0.0.1 is synced on DFS volume /tmp/storage199463140 for file /tmp/storage199463140/monitor/127.0.0.1 (since modified: 6.667821ms)
I0124 11:18:24.416925 17580 monitor_test.go:45] ==== received remoteIP:127.0.0.1 isUpdated:true
I0124 11:18:25.416987 17580 monitor.go:45] writing DFS file /tmp/storage199463140/monitor/127.0.0.1
I0124 11:18:26.417365 17580 monitor.go:45] writing DFS file /tmp/storage199463140/monitor/127.0.0.1
I0124 11:18:27.417135 17580 monitor.go:145] file /tmp/storage199463140/monitor/127.0.0.1 was modified 1.002913479s ago
I0124 11:18:27.417180 17580 monitor.go:117] remote agent 127.0.0.1 is synced on DFS volume /tmp/storage199463140 for file /tmp/storage199463140/monitor/127.0.0.1 (since modified: 1.002913479s)
I0124 11:18:27.417206 17580 monitor_test.go:45] ==== received remoteIP:127.0.0.1 isUpdated:true
I0124 11:18:27.417840 17580 monitor.go:45] writing DFS file /tmp/storage199463140/monitor/127.0.0.1
I0124 11:18:28.418122 17580 monitor.go:45] writing DFS file /tmp/storage199463140/monitor/127.0.0.1
I0124 11:18:29.418456 17580 monitor.go:45] writing DFS file /tmp/storage199463140/monitor/127.0.0.1
I0124 11:18:30.417380 17580 monitor.go:145] file /tmp/storage199463140/monitor/127.0.0.1 was modified 1.00315692s ago
I0124 11:18:30.417426 17580 monitor.go:117] remote agent 127.0.0.1 is synced on DFS volume /tmp/storage199463140 for file /tmp/storage199463140/monitor/127.0.0.1 (since modified: 1.00315692s)
I0124 11:18:30.417440 17580 monitor_test.go:45] ==== received remoteIP:127.0.0.1 isUpdated:true
I0124 11:18:30.418768 17580 monitor.go:45] writing DFS file /tmp/storage199463140/monitor/127.0.0.1
I0124 11:18:31.419084 17580 monitor.go:45] writing DFS file /tmp/storage199463140/monitor/127.0.0.1
I0124 11:18:32.419441 17580 monitor.go:45] writing DFS file /tmp/storage199463140/monitor/127.0.0.1
I0124 11:18:33.416525 17580 monitor_test.go:75] ==== stopping writer
I0124 11:18:33.416573 17580 monitor.go:55] no longer writing remote monitor status for DFS volume /tmp/storage199463140 to /tmp/storage199463140/monitor/127.0.0.1
I0124 11:18:33.417602 17580 monitor.go:145] file /tmp/storage199463140/monitor/127.0.0.1 was modified 1.003377201s ago
I0124 11:18:33.417631 17580 monitor.go:117] remote agent 127.0.0.1 is synced on DFS volume /tmp/storage199463140 for file /tmp/storage199463140/monitor/127.0.0.1 (since modified: 1.003377201s)
I0124 11:18:33.417644 17580 monitor_test.go:45] ==== received remoteIP:127.0.0.1 isUpdated:true
I0124 11:18:36.417823 17580 monitor.go:145] file /tmp/storage199463140/monitor/127.0.0.1 was modified 4.003597666s ago
W0124 11:18:36.417887 17580 monitor.go:115] remote agent 127.0.0.1 is not synced DFS volume /tmp/storage199463140 for file /tmp/storage199463140/monitor/127.0.0.1 (since modified: 4.003597666s)
I0124 11:18:36.417921 17580 monitor_test.go:45] ==== received remoteIP:127.0.0.1 isUpdated:false
I0124 11:18:39.418101 17580 monitor.go:145] file /tmp/storage199463140/monitor/127.0.0.1 was modified 7.003875619s ago
W0124 11:18:39.418145 17580 monitor.go:115] remote agent 127.0.0.1 is not synced DFS volume /tmp/storage199463140 for file /tmp/storage199463140/monitor/127.0.0.1 (since modified: 7.003875619s)
I0124 11:18:39.418160 17580 monitor_test.go:45] ==== received remoteIP:127.0.0.1 isUpdated:false
I0124 11:18:42.418304 17580 monitor.go:145] file /tmp/storage199463140/monitor/127.0.0.1 was modified 10.004078255s ago
W0124 11:18:42.418346 17580 monitor.go:115] remote agent 127.0.0.1 is not synced DFS volume /tmp/storage199463140 for file /tmp/storage199463140/monitor/127.0.0.1 (since modified: 10.004078255s)
I0124 11:18:42.418360 17580 monitor_test.go:45] ==== received remoteIP:127.0.0.1 isUpdated:false
I0124 11:18:45.416935 17580 monitor_test.go:89] ==== current updatedCount:0
I0124 11:18:45.416995 17580 monitor.go:130] no longer monitoring status for DFS volume /tmp/storage199463140
--- PASS: TestMonitorVolume (24.00 seconds)
=== RUN TestServer
--- SKIP: TestServer (0.00 seconds)
	server_test.go:60: 
PASS
ok  	github.com/control-center/serviced/coordinator/storage	24.010s
```

### CODE REVIEW

File: coordinator/storage/monitor.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+func init() {
+	monitoredHosts = make(map[string]bool)
+}
+
+// UpdateRemoteMonitorFile writes a tiny file to the DFS volume at given cycle
+func UpdateRemoteMonitorFile(localPath string, writeInterval time.Duration, ipAddr string) {
+	glog.Infof("writing remote monitor file to DFS volume %s at interval: %s", localPath, writeInterval)
+
+	for {
+		remoteFile := path.Join(localPath, monitorSubDir, ipAddr)
```
Comment (alapidas) [inline coordinator/storage/monitor.go]: Can this line be moved outside the for loop?

Comment (wplu) [inline coordinator/storage/monitor.go]: yes

File: coordinator/storage/monitor.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+				glog.Warningf("remote agent is not synced on DFS volume %s for file %s: %s", mountpoint, remoteFile, err)
+			} else if sinceModified > monitorInterval {
+				glog.Warningf("remote agent is not synced DFS volume %s for file %s (since modified: %s)", mountpoint, remoteFile, sinceModified)
+			} else {
+				glog.Infof("remote agent is synced on DFS volume %s for file %s (since modified: %s)", mountpoint, remoteFile, sinceModified)
+				hasUpdatedFile = true
+			}
+
+			SetMonitorRemoteHost(remoteIP, hasUpdatedFile)
+			pollUpdateFunc(mountpoint, hasUpdatedFile)
```
Comment (alapidas) [inline coordinator/storage/monitor.go]: What func will this be?  What will it do?

File: coordinator/storage/monitor.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+//
+func AddMonitorRemoteHost(ipAddr string) {
+	monitoredHostsLock.Lock()
+	defer monitoredHostsLock.Unlock()
+	monitoredHosts[ipAddr] = true
+}
+
+//
+func SetMonitorRemoteHost(ipAddr string, hasUpdated bool) {
```
Comment (alapidas) [inline coordinator/storage/monitor.go]: I can't see anywhere where we actually use the bool part of an entry in this map - am I missing something?

File: coordinator/storage/monitor.go (hunk 2 above)
Comment (wplu) [inline coordinator/storage/monitor.go]: see server.go

File: coordinator/storage/monitor.go (hunk 3 above)
Comment (wplu) [inline coordinator/storage/monitor.go]: the bool can be used to determine if a status change occurred - unused for now

File: coordinator/storage/client.go [hunk 4]
```diff
@@ -155,9 +155,11 @@ func (c *Client) loop() {
 				if err == nfs.ErrNfsMountingUnsupported {
 					glog.Errorf("install the nfs-common package: %s", err)
 				}
-				glog.Errorf("problem mouting %s: %s", leaderNode.ExportPath, err)
+				glog.Errorf("problem mounting %s: %s", leaderNode.ExportPath, err)
 				continue
 			}
+
+			go UpdateRemoteMonitorFile(c.localPath, getDefaultDFSMonitorRemoteInterval(), c.host.IPAddr)
```
Comment (alapidas) [inline coordinator/storage/client.go]: Does this guy need the c.closing channel passed in?

**Comment by alapidas** [comment]
retest this please

File: coordinator/storage/client.go (hunk 4 above)
Comment (wplu) [inline coordinator/storage/client.go]: yes it does - in the next commit

**Comment by wplu** [comment]
TODO: change most of those debugging glog.V(0) to V(4)

**Comment by wplu** [comment]
TODO: remove hosts when they are not active

10.87.209.168  # master
10.87.209.184
10.87.209.228

seen when remote hosts are down:
```
I0124 17:47:54.378192 02143 monitor.go:140] unable to stat file: /exports/serviced_var/monitor/10.87.209.228 stat /exports/serviced_var/monitor/10.87.209.228: no such file or directory
W0124 17:47:54.378247 02143 monitor.go:113] remote agent 10.87.209.228 is not synced on DFS volume /exports/serviced_var for file /exports/serviced_var/monitor/10.87.209.228: stat /exports/serviced_var/monitor/10.87.209.228: no such file or directory
W0124 17:47:54.378262 02143 server.go:80] DFS NFS volume /exports/serviced_var may be unexported - further action may be needed i.e: restart nfs
W0124 17:47:54.378268 02143 server.go:82] Not restarting NFS service due to configuration setting: SERVICED_MONITOR_DFS_MASTER_RESTART=0
I0124 17:47:54.378278 02143 monitor.go:140] unable to stat file: /exports/serviced_var/monitor/10.87.209.168 stat /exports/serviced_var/monitor/10.87.209.168: no such file or directory
W0124 17:47:54.378284 02143 monitor.go:113] remote agent 10.87.209.168 is not synced on DFS volume /exports/serviced_var for file /exports/serviced_var/monitor/10.87.209.168: stat /exports/serviced_var/monitor/10.87.209.168: no such file or directory
W0124 17:47:54.378290 02143 server.go:80] DFS NFS volume /exports/serviced_var may be unexported - further action may be needed i.e: restart nfs
W0124 17:47:54.378294 02143 server.go:82] Not restarting NFS service due to configuration setting: SERVICED_MONITOR_DFS_MASTER_RESTART=0
I0124 17:47:54.378300 02143 monitor.go:140] unable to stat file: /exports/serviced_var/monitor/10.87.209.184 stat /exports/serviced_var/monitor/10.87.209.184: no such file or directory
W0124 17:47:54.378306 02143 monitor.go:113] remote agent 10.87.209.184 is not synced on DFS volume /exports/serviced_var for file /exports/serviced_var/monitor/10.87.209.184: stat /exports/serviced_var/monitor/10.87.209.184: no such file or directory
W0124 17:47:54.378311 02143 server.go:80] DFS NFS volume /exports/serviced_var may be unexported - further action may be needed i.e: restart nfs
W0124 17:47:54.378315 02143 server.go:82] Not restarting NFS service due to configuration setting: SERVICED_MONITOR_DFS_MASTER_RESTART=0
```

**Comment by wplu** [comment]
GOOD - remote agent is writing to correct files at the interval specified:
```
I0124 17:52:29.802471 19933 monitor.go:45] writing DFS file /opt/serviced/var/monitor/10.87.209.228
I0124 17:53:29.804102 19933 monitor.go:45] writing DFS file /opt/serviced/var/monitor/10.87.209.228
I0124 17:54:29.805562 19933 monitor.go:45] writing DFS file /opt/serviced/var/monitor/10.87.209.228
```

GOOD - the master sees it:
```
I0124 17:53:54.378981 02143 monitor.go:145] file /exports/serviced_var/monitor/10.87.209.228 was modified 24.571697218s ago
I0124 17:53:54.379031 02143 monitor.go:117] remote agent 10.87.209.228 is synced on DFS volume /exports/serviced_var for file /exports/serviced_var/monitor/10.87.209.228 (since modified: 24.571697218s)

I0124 17:53:54.379049 02143 monitor.go:140] unable to stat file: /exports/serviced_var/monitor/10.87.209.168 stat /exports/serviced_var/monitor/10.87.209.168: no such file or directory
W0124 17:53:54.379061 02143 monitor.go:113] remote agent 10.87.209.168 is not synced on DFS volume /exports/serviced_var for file /exports/serviced_var/monitor/10.87.209.168: stat /exports/serviced_var/monitor/10.87.209.168: no such file or directory
W0124 17:53:54.379068 02143 server.go:80] DFS NFS volume /exports/serviced_var may be unexported - further action may be needed i.e: restart nfs
W0124 17:53:54.379073 02143 server.go:82] Not restarting NFS service due to configuration setting: SERVICED_MONITOR_DFS_MASTER_RESTART=0
I0124 17:53:54.379079 02143 monitor.go:140] unable to stat file: /exports/serviced_var/monitor/10.87.209.184 stat /exports/serviced_var/monitor/10.87.209.184: no such file or directory
W0124 17:53:54.379085 02143 monitor.go:113] remote agent 10.87.209.184 is not synced on DFS volume /exports/serviced_var for file /exports/serviced_var/monitor/10.87.209.184: stat /exports/serviced_var/monitor/10.87.209.184: no such file or directory
W0124 17:53:54.379090 02143 server.go:80] DFS NFS volume /exports/serviced_var may be unexported - further action may be needed i.e: restart nfs
W0124 17:53:54.379094 02143 server.go:82] Not restarting NFS service due to configuration setting: SERVICED_MONITOR_DFS_MASTER_RESTART=0
```

**Comment by wplu** [comment]
TODO: need to start a monitor writer on the master agent

**Comment by alapidas** [comment]
This is good to go!  May want to squash first though.

**Comment by wplu** [comment]
retest this please

**Comment by wplu** [comment]
retest this please

**Comment by wplu** [comment]
DEMO:
```
I0126 05:20:57.286069 27137 monitor.go:150] monitoring DFS export info for DFS volume /exports/serviced_var at polling interval: 2m0s
I0126 05:20:57.320713 27137 nfs.go:76] started nfs server:


I0126 05:20:57.364806 27137 daemon.go:584] Waiting for Storage Leader
I0126 05:20:57.364840 27137 monitor.go:64] updating DFS volume monitor file /opt/serviced/var/monitor/10.87.209.168 at write interval: 1m0s
I0126 05:20:57.364901 27137 client.go:213] updating DFS remote client UpdatedAt for node /storage/clients/10.87.209.168 at interval 1m0s


I0126 05:24:57.289392 27137 monitor.go:305] DFS not monitoring non-active host 10.87.209.228:  HostID:570ae4d1  UpdatedAt:2015-01-26 05:21:12.914979276 +0000 UTC  lastseen:3m44.37397201s ago
I0126 05:24:57.290337 27137 monitor.go:175] DFS active remotes has changed to: [10.87.209.168 10.87.209.184]
```

**Pull request merged**